client/src/controllers/SwapController.ts

```diff
@@ -262,38 +275,113 @@ export class SwapController extends Controller<
     }) as CustomEvent<{ requestUrl: string }>;
 
     if (beginEvent.defaultPrevented) return Promise.resolve();
-
+    const formMethod = this.formElement.getAttribute('method') || undefined;
     return fetch(requestUrl, {
-      headers: { 'x-requested-with': 'XMLHttpRequest' },
+      headers: {
+        'x-requested-with': 'XMLHttpRequest',
+        [WAGTAIL_CONFIG.CSRF_HEADER_NAME]: WAGTAIL_CONFIG.CSRF_TOKEN,
+      },
       signal,
+      method: formMethod,
+      body: formMethod !== 'get' ? data : undefined,
     })
-      .then((response) => {
+      .then(async (response) => {
         if (!response.ok) {
           throw new Error(`HTTP error! Status: ${response.status}`);
         }
+        if (this.jsonPathValue) {
+          let html: unknown;
+          try {
+            const json: Record<string, unknown> = await response.json();
+
+            this.dispatch('json', {
+              cancelable: false,
+              detail: { requestUrl, data: json },
+            });
+
+            html = this.jsonPathValue
+              .split('.')
+              .reduce<unknown>(
+                (acc, key) => (acc as Record<string, unknown>)[key],
+                json,
+              );
+          } catch {
+            html = undefined;
+          }
+
+          if (typeof html !== 'string') {
+            throw new Error(
+              `Unable to parse as JSON at path "${this.jsonPathValue}" to a string`,
+            );
+          }
+          return html;
+        }
         return response.text();
       })
       .then((results) => {
-        target.innerHTML = results;
+        const write = async () => {
+          target.innerHTML = results;
+
+          if (this.reflectValue) {
+            const event = this.dispatch('reflect', {
+              cancelable: true,
+              detail: { requestUrl },
+              target,
+            });
+            if (!event.defaultPrevented) {
+              this.reflectParams(requestUrl);
+            }
+          }
 
-        if (this.reflectValue) {
-          const event = this.dispatch('reflect', {
-            cancelable: true,
-            detail: { requestUrl },
+          this.dispatch('success', {
+            cancelable: false,
+            detail: { requestUrl, results },
             target,
           });
-          if (!event.defaultPrevented) {
-            this.reflectParams(requestUrl);
-          }
-        }
 
-        this.dispatch('success', {
-          cancelable: false,
-          detail: { requestUrl, results },
-          target,
-        });
+          return results;
+        };
+
+        // If the currently focused element is within the target container,
+        // defer the write until the focus has left the container.
```

Comment:
Upon further testing, I noticed this affects tooltips more than we discussed:

- Hovering over the open tooltip (for example to click a button within) results in it disappearing 
- "Clicking" the tooltip and then moving focus away also results in it disappearing

Is it possible to better manage those scenarios?

Reply:
I've added a commit that handles the case when there are tooltips shown for an element in the target container